File: Typescript/src/clope/RuleSet.ts
import { ColumnNumber, TransactionElement } from "../common/Typings";

export default class RuleSet<T> {
    private ConvertFunc: (row: T) => TransactionElement[];
    private skipElements: Set<TransactionElement>;
    private columnsToSkip: Set<ColumnNumber>;

    constructor(config: {
        ConvertFunc: ((row: T) => TransactionElement[])
        nullElements?: TransactionElement[],
        columnsToSkip?: number[],
    }) {
        this.skipElements = this.CreateSet(config.nullElements);
        this.columnsToSkip = this.CreateSet(config.columnsToSkip);
        this.ConvertFunc = config.ConvertFunc;
    }

    public Update(config: {
        ConvertFunc: ((row: T) => TransactionElement[])
        nullElements?: TransactionElement[],
        columnsToSkip?: ColumnNumber[],
    }): void {
        this.skipElements = this.CreateSet(config.nullElements);
        this.columnsToSkip = this.CreateSet(config.columnsToSkip);
        this.ConvertFunc = config.ConvertFunc;
    }

    public Apply(row: T): TransactionElement[] {
        return this.ConvertFunc(row);
    }

    public ApplyWithRules(row: T): TransactionElement[] {
        const filteredElements = this.Apply(row);
        for (let columnNumber = 0; columnNumber < filteredElements.length; columnNumber++) {
            const element = filteredElements[columnNumber];
            if (this.columnsToSkip.has(columnNumber) || this.skipElements.has(element)) {
                filteredElements[columnNumber] = null;
            }
        }
        return filteredElements;
    }

    private CreateSet<TArray>(array?: TArray[]) {
        return array ? new Set(array) : new Set();
    }
}

File: c#/Clope/clope/RuleSet.cs
using System;
using System.Collections.Generic;
using Clope.common;

namespace Clope.clope
{
    public struct RuleConfig<T>
    {
        public Func<T, object[]> ConvertFunc;
        public HashSet<object> skipElements;
        public HashSet<int> columnsToSkip;
    }
    public class RuleSet<T>
    {
        //protected List<object> ConvertFunc(T row);
        private RuleConfig<T> _config;

        public RuleSet(Func<T, object[]> convertFunc,
                        object[] nullElements,
                        int[] columnsToSkip)
        {
            _config.ConvertFunc = convertFunc;
            _config.skipElements = CreateSet(nullElements);
            _config.columnsToSkip = CreateSet(columnsToSkip);
        }

        public void Update(Func<T, object[]> convertFunc,
                        object[] nullElements,
                        int[] columnsToSkip)
        {
            _config.ConvertFunc = convertFunc;
            _config.skipElements = CreateSet(nullElements);
            _config.columnsToSkip = CreateSet(columnsToSkip);
        }

        public object[] ApplyWithRules(T row)
        {
            var filteredElements = Apply(row);
            for (var columnNumber = 0; columnNumber < filteredElements.Length; columnNumber++)
            {
                var element = filteredElements[columnNumber];
                if (_config.columnsToSkip.Contains(columnNumber) || _config.skipElements.Contains(element))
                {
                    filteredElements[columnNumber] = null;
                }
            }
            return filteredElements;
        }

        public object[] Apply(T row)
        {
            return _config.ConvertFunc(row);
        }

        private static HashSet<TArray> CreateSet<TArray>(TArray[] array)
        {
            return (array != null) ? new HashSet<TArray>(array) : new HashSet<TArray>();
        }
    }
}